### PR TITLE
Add video input support for vision-language models

### PR DESCRIPTION
**Pull request by stikves**
## Summary

- Video frame extraction pipeline using `AVAssetImageGenerator` with uniform and FPS-based sampling strategies
- `VideoInput` delivers frames lazily via a concrete `VideoFrameSequence` (`AsyncSequence` + `Sendable`) to keep peak memory at 1–2 decoded frames
- VLM engine encodes each frame through the existing vision encoder and concatenates embeddings along the sequence dimension
- `--video`, `--video-frames`, `--video-sampling` flags on `llm-runner`
- `VisionConfig` extended with optional `max_video_frames` and `tokens_per_frame` fields (backwards compatible)
- VLM export gains `--num-frames` for native multi-frame temporal encoding (`grid_t > 1`)
- Vectorized RGBA→CHW transpose in `ImagePreprocessor` using `vDSP_vsadd` stride-4 gather

## New types

| Type | Module | Role |
|------|--------|------|
| `FrameSamplingStrategy` | CoreAIShared | `.uniform(count:)` / `.fps(rate:, maxFrames:)` |
| `VideoInput` | CoreAIShared | Frame source with lazy `VideoFrameSequence` |
| `VideoFrameSequence` | CoreAIShared | Concrete `AsyncSequence & Sendable` over `VideoFrame` |
| `VideoFrame` | CoreAIShared | `CGImage` + timestamp |

## Usage

```bash
swift run -c release llm-runner \
    --model exports/qwen3_vl_2b.llmasset \
    --video path/to/video.mp4 \
    --video-frames 8 \
    --prompt "Describe what happens in this video"
```

## Test plan

- [x] `FrameSamplingStrategy`: uniform/fps math, edge cases (single frame, clamp, zero duration, negative duration)
- [x] `VideoInput.fromFrames`: wraps `[CGImage]`, empty array
- [x] `VisionConfig`: decodes with/without video fields (backwards compat)
- [x] `ImagePreprocessor`: existing tests pass with vectorized transpose
- [x] End-to-end: Qwen3-VL-2B + screen recording (4 frames, 784 visual tokens)

### CODE REVIEW

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAISequentialEngine.swift [hunk 1]
```diff
@@ -475,84 +475,7 @@ public final class CoreAISequentialEngine: InferenceEngine, @unchecked Sendable
         cleanupSpan.end()
     }
 
-    // MARK: - KV Cache (dynamic growth)
```
Comment (stikves) [inline swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAISequentialEngine.swift]: These fixes cherry picked from https://github.com/apple/coreai-models/pull/142

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIPipelinedEngine.swift [hunk 2]
```diff
@@ -1141,7 +1140,6 @@ private struct EngineImpl: ~Copyable {
         var asyncStates = InferenceFunction.AsyncMutableViews()
         asyncStates.insert(&keyState, for: keyCacheName)
         asyncStates.insert(&valState, for: valueCacheName)
-        additionalStates?.insertAll(into: &asyncStates)
```
Comment (kevchengcodes) [inline swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIPipelinedEngine.swift]: are these deletions also from #142?

File: swift/Sources/CoreAIShared/Video/FrameSamplingStrategy.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import Foundation
+
+/// How to extract frames from a video for vision encoding.
+public enum FrameSamplingStrategy: Sendable {
+    /// N evenly-spaced frames across the video duration.
+    case uniform(count: Int)
+    /// One frame every `rate` seconds, capped at `maxFrames`.
+    case fps(rate: Double, maxFrames: Int)
+
+    /// Compute frame timestamps for a video with the given duration and frame rate.
```
Comment (kevchengcodes) [inline swift/Sources/CoreAIShared/Video/FrameSamplingStrategy.swift]: maybe add here that this is calculating the timestamp in the middle of the frame

File: swift/Sources/CoreAIShared/Video/VideoInput.swift [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    }
+}
+
+/// Extracted video frames ready for vision encoding.
+///
+/// Frames are delivered lazily via `VideoFrameSequence`. The engine processes
+/// and releases each frame incrementally, keeping peak memory at 1-2
+/// decoded frames plus accumulated embeddings.
+public struct VideoInput: Sendable {
+    /// Number of frames, if known ahead of time (nil for live streams).
```
Comment (kevchengcodes) [inline swift/Sources/CoreAIShared/Video/VideoInput.swift:51]: Do we support live streams with this PR?

File: swift/Sources/CoreAIShared/Video/VideoInput.swift [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        self.frameCount = frameCount
+        self.duration = duration
+        self.frames = frames
+    }
+
+    /// Extract frames from a local video file.
+    ///
+    /// - Parameters:
+    ///   - url: File URL to a video (MP4, MOV, etc.).
+    ///   - sampling: Frame sampling strategy (default: 8 uniform frames).
```
Comment (kevchengcodes) [inline swift/Sources/CoreAIShared/Video/VideoInput.swift:72]: this is an important default that seems a little hidden. Is there a way to make this more visible, or maybe we don't provide a default?

File: swift/Sources/CoreAIShared/Video/VideoInput.swift [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    ///
+    /// Timestamps are assigned as 0-based indices since the source
+    /// may not have meaningful timing information.
+    public static func fromFrames(
+        _ images: [CGImage]
+    ) -> VideoInput {
+        let count = images.count
+        let stream = AsyncThrowingStream<VideoFrame, Error> { continuation in
+            for (i, image) in images.enumerated() {
+                continuation.yield(VideoFrame(image: image, timestamp: Double(i)))
```
Comment (kevchengcodes) [inline swift/Sources/CoreAIShared/Video/VideoInput.swift]: does it make sense to call them timestamps in this case?

File: swift/Sources/CoreAIShared/Video/VideoInput.swift [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        let (count, duration, frames) = try await VideoFrameExtractor.extractFrames(
+            from: url, sampling: sampling, skipErrors: skipErrors)
+        return VideoInput(frameCount: count, duration: duration, frames: frames)
+    }
+
+    /// Wrap pre-extracted frames (e.g. from camera capture).
+    ///
+    /// Timestamps are assigned as 0-based indices since the source
+    /// may not have meaningful timing information.
+    public static func fromFrames(
```
Comment (kevchengcodes) [inline swift/Sources/CoreAIShared/Video/VideoInput.swift:87]: seems unused

File: swift/Sources/Tools/llm-runner/LLMRunnerMain.swift [hunk 8]
```diff
@@ -50,6 +50,11 @@ enum ImageInfoMode: String, ExpressibleByArgument, Sendable {
     case auto
 }
 
+enum VideoSamplingMode: String, ExpressibleByArgument, Sendable {
+    case uniform
+    case fps
+}
+
```
Comment (kevchengcodes) [inline swift/Sources/Tools/llm-runner/LLMRunnerMain.swift:62]: can we de-dupe this with `FrameSamplingStrategy`? I doubt we'll be adding to it often, but it'll be tougher to maintain both enums.

File: swift/Sources/Tools/llm-runner/LLMRunnerMain.swift [hunk 9]
```diff
@@ -903,15 +942,96 @@ struct LLMRunner: AsyncParsableCommand, Sendable {
...
+        guard let visionConfig = bundle.visionConfig else {
+            print("Error: VLM bundle missing 'vision' config in metadata.json")
+            throw ExitCode.failure
+        }
+
+        let sampling: FrameSamplingStrategy
+        switch videoSampling {
+        case .fps:
+            sampling = .fps(rate: 1.0, maxFrames: videoFrames)
```
Comment (kevchengcodes) [inline swift/Sources/Tools/llm-runner/LLMRunnerMain.swift]: will users ever want to change the FPS rate, or will it always be 1.0?

File: swift/Sources/Tools/llm-runner/LLMRunnerMain.swift [hunk 10]
```diff
@@ -903,15 +942,96 @@ struct LLMRunner: AsyncParsableCommand, Sendable {
...
+        case .uniform:
+            sampling = .uniform(count: videoFrames)
+        }
+
+        if !CLILogger.isVerbose {
+            print("Generating...")
+        }
```
Comment (kevchengcodes) [inline swift/Sources/Tools/llm-runner/LLMRunnerMain.swift:996]: why only print this when it is NOT verbose?

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIPipelinedEngine.swift (hunk 2 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIPipelinedEngine.swift]: Yes, will be removed after rebase

File: swift/Sources/CoreAIShared/Video/VideoInput.swift (hunk 4 above)
Comment (stikves) [reply swift/Sources/CoreAIShared/Video/VideoInput.swift:51]: Not yet, this builds up the foundation

File: swift/Sources/Tools/llm-runner/LLMRunnerMain.swift (hunk 10 above)
Comment (stikves) [reply swift/Sources/Tools/llm-runner/LLMRunnerMain.swift:996]: In verbose we already print detailed information.

File: swift/Sources/CoreAIShared/Video/VideoInput.swift (hunk 7 above)
Comment (stikves) [reply swift/Sources/CoreAIShared/Video/VideoInput.swift:87]: Used in tests (also left as part of public API)

File: swift/Sources/CoreAIShared/Video/VideoInput.swift (hunk 6 above)
Comment (stikves) [reply swift/Sources/CoreAIShared/Video/VideoInput.swift]: Moving to the `index` naming

File: swift/Sources/Tools/llm-runner/LLMRunnerMain.swift (hunk 8 above)
Comment (stikves) [reply swift/Sources/Tools/llm-runner/LLMRunnerMain.swift:62]: Makes sense

File: swift/Sources/CoreAIShared/Video/VideoInput.swift (hunk 5 above)
Comment (stikves) [reply swift/Sources/CoreAIShared/Video/VideoInput.swift:72]: 8 seems to be the common strategy for the models I have seen so far. When we need to change, we can design a mechanism to update.

File: swift/Sources/Tools/llm-runner/LLMRunnerMain.swift (hunk 9 above)
Comment (stikves) [reply swift/Sources/Tools/llm-runner/LLMRunnerMain.swift]: same, 1.0 seems standard for VLMs, but we can think about adding a CLI flag later (don't want to prematurely add the option without actually adhering to it correctly)

**Review by tjia1818 — APPROVED** [review]
